src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java

```diff
@@ -0,0 +1,126 @@
+
+        MysqlSQLXML xmlTest = new MysqlSQLXML(null, connProperties);
+
+        assertTrue(xmlTest.getAllowXmlUnsafeExternalEntity());
+
+    }
+
+    @Test
+    public void testXXEInjectionPrevention() throws SQLException {
+
+        final String xmlString = "<?xml version=\"1.0\" encoding=\"ISO-8859-1\"?>\n" +
+            "<!DOCTYPE foo [\n" +
+            "<!ELEMENT foo ANY >\n" +
+            "<!ENTITY xxe SYSTEM \"file:///dev/nonExistantFile\" >]><foo>&xxe;</foo>";
+        final String url =
+            "jdbc:mysql://somehost:1234/test";
+        final ConnectionUrl connUrl = ConnectionUrl.getConnectionUrlInstance(url, new Properties());
+        final String expectedErrorMsg = "DOCTYPE is disallowed when the feature \"http://apache.org/xml/features/disallow-doctype-decl\" set to true.";
+
+        JdbcPropertySetImpl connProperties = new JdbcPropertySetImpl();
+        connProperties.initializeProperties(connUrl.getMainHost().exposeAsProperties());
+
+        MysqlSQLXML xmlTest = new MysqlSQLXML(null, connProperties);
+
+        xmlTest.setString(xmlString);
+        SQLException exception = assertThrows( SQLException.class, () -> xmlTest.getSource(DOMSource.class));
+
+        // Assert correct error code was produced
+        assertSame("S1009", exception.getSQLState());
+        // Assert that the reason for this exception was because of the security measurements put in place (DTDs disabled)
+        assertTrue(expectedErrorMsg.equals(exception.getMessage()));
+    }
+
+    @Test
+    public void testXXEInjectionAllowance() throws SQLException {
+
+        final String xmlString = "<?xml version=\"1.0\" encoding=\"ISO-8859-1\"?>\n" +
+                "<!DOCTYPE foo [\n" +
+                "<!ELEMENT foo ANY >\n" +
+                "<!ENTITY xxe SYSTEM \"file:///dev/nonExistantFile\" >]><foo>&xxe;</foo>";
+        final String url =
+                "jdbc:mysql://somehost:1234/test?"
+                        + PropertyKey.allowXmlUnsafeExternalEntity.getKeyName()
+                        + "=true";
+        final String expectedErrorMsg = "/dev/nonExistantFile (No such file or directory)";
+        final ConnectionUrl connUrl = ConnectionUrl.getConnectionUrlInstance(url, new Properties());
+        JdbcPropertySetImpl connProperties = new JdbcPropertySetImpl();
+        connProperties.initializeProperties(connUrl.getMainHost().exposeAsProperties());
+
+        MysqlSQLXML xmlTest = new MysqlSQLXML(null, connProperties);
+
+        xmlTest.setString(xmlString);
+        SQLException exception = assertThrows( SQLException.class, () -> xmlTest.getSource(DOMSource.class));
+
+        // Assert correct error code was produced
+        assertSame("S1009", exception.getSQLState());
+        // Assert that the reason for this exception is because the file does not exist (and not because DTDs is disabled)
+        assertTrue(expectedErrorMsg.equals(exception.getMessage()));
```

Comment:
It fails in Windows. Exception message is
\dev\nonExistantFile (The system cannot find the path specified)

It's probably better to check for an expected substring in exception message.

Reply:
I just pushed a changed but just realized even the text is completely different.

Reply:
Changed it to check if the string contains either "(The system cannot find the path specified)" or "(No such file or directory)"

Reply:
Another option we could do is to check if the error message is different than the previous test.